text/0000-constants-in-patterns.md

```diff
@@ -0,0 +1,306 @@
+
+Range patterns are only allowed on integers, `char`, and floats; for floats, neither end must be a `NaN`.
+
+The *behavior* of such a constant as a pattern is the same as the corresponding native pattern.
+On floats are raw pointers, pattern matching behaves like `==`,
+which means in particular that the value `-0.0` matches the pattern `0.0`, and NaN values match no pattern.
+
+## Breaking changes
+
+This RFC breaks code that compiles today, but only code that already emits a future compatibility lint:
+- Matching on constants that do not implement `PartialEq` sometimes accidentally works, but triggers `const_patterns_without_partial_eq`.
+  This lint landed with Rust 1.74 (the most recent stable release), and is shown in dependencies as well.
+- Matching on `struct`/`enum` that do not `derive(PartialEq)` is accidentally possible under some conditions, but triggers `indirect_structural_match`.
+  This has been a future-compatibility lint for many years, though it is currently not shown in dependencies.
+- Matching on function pointers, or raw pointers that are not defined as a constant integer, triggers `pointer_structural_match`.
+  This only recently landed (Rust 1.75, currently in beta), and is not currently shown in dependencies.
+  Crater found [three cases](https://github.com/rust-lang/rust/pull/116930#issuecomment-1784648989) across the ecosystem where `match` was used to compare function pointers;
+  that code is buggy for the reasons mentioned above that make comparing function pointers unreliable.
+- Matching on floats triggers `illegal_floating_point_literal_pattern`. This triggers on *all* float matches, not just the ones forbidden by this RFC.
+  It has been around for years, but is not currently shown in dependencies.
+
+When the RFC gets accepted, the floating-point lint should be adjusted to only cover the cases we are really going to reject,
+and all of them should be shown in dependencies or directly turned into hard errors.
+
+## Compiler/library cleanup
+
+There also exists the `nontrivial_structural_match` future compatibility lint;
+it is not needed for this RFC so it can be removed when the RFC gets accepted.
+
+Similarly, the `StructuralEq` trait no longer serves a purpose and can be removed.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- The biggest drawback of this proposal is that it conflates `derive(PartialEq)` with a semver-stable promise that this type will always have structural equality.
+  Once a type has `derive(PartialEq)`, it may appear in patterns, so replacing this `PartialEq` by a custom implementation is a breaking change.
+  Once the `StructuralPartialEq` trait is stable, `derive(PartialEq)` *can* be replaced by a custom implementation as long as one also implements `StructuralPartialEq`,
+  but that entails a promise that the `impl` still behaves structurally, including on all private fields.
```

Comment:
(It would be breaking, which means we almost certainly shouldn't do it, but) Given the logic that a `const` pattern is “logically desugared” into the corresponding literal pattern, it could be justified in forbidding matching over any type with inaccessible fields (or `#[non_exhaustive]`).

(Fun: because this sees into private fields, you can exhaustively match a type with all private fields. I unfortunately don't have my set up to test multiple crates right now, but I will be *extremely* amused if it turns out you can exhaustively match a `#[non_exhaustive]` struct. Downright dtolnay/rust-quiz material.)

Reply:
> it could be justified in forbidding matching over any type with inaccessible fields (or #[non_exhaustive]).

This is somewhat deliberate I think: a library can choose to expose certain constants that people can match on despite there being private fields.

I don't think this will let you do exhaustive matching on non_exhaustive enums. For structs there's no separate notion of exhaustiveness anyway, it's all per-field?

Reply:
For structs and struct variants, `#[non_exhaustive]` means "I may add fields in the future so you must always deconstruct with `, ..`". That doesn't change anything for constants-interpreted-as-patterns. I can't find a test for that case though, I'll add one when I get the chance

Reply:
> I will be _extremely_ amused if it turns out you can exhaustively match a `#[non_exhaustive]` struct.

You can :facepalm: 
```rust
# bar/src/lib.rs
#[non_exhaustive]
#[derive(PartialEq, Eq)]
pub struct Bar(bool);
pub const TRUE: Bar = Bar(true);
pub const FALSE: Bar = Bar(false);
```
```rust
# foo/src/main.rs
fn main() {
    match bar::TRUE {
        bar::TRUE => {}
        bar::FALSE => {}
    }
}
```
```console
> cargo check
    Checking bar v0.1.0 (/tmp/scratch.rust.2023-12-17T12-41.GFlTlw/foo/bar)
    Checking foo v0.1.0 (/tmp/scratch.rust.2023-12-17T12-41.GFlTlw/foo)
    Finished dev [unoptimized + debuginfo] target(s) in 0.11s
```

Reply:
That doesn't seem surprising to me, since this does not deconstruct the value at all, it just compares for equality.

But if this should lint/error then please file an issue.

Reply:
I don't think that should error. `#[non_exhaustive]` means "adding fields musn't be a breaking change", which is the case when using constants. The tricky case is accessing private fields, which is addressed in your RFC.

Reply:
> I don't think that should error. `#[non_exhaustive]` means "adding fields musn't be a breaking change", which is the case when using constants. The tricky case is accessing private fields, which is addressed in your RFC.

How would you add a non-unit type field without breaking the match?
Adding such fields would add more possible states, that would not be coverable by the two existing const patterns.

Reply:
Oh hm fair point. Maybe it should be opt-in like for private fields

Reply:
This appears to be a new (at least to me) semver hazard. I think it'd be very surprising to maintainers that it's possible to cause a breaking change by adding/tweaking a private field on a `#[non_exhaustive]` type.

I think this should either be a future-incompat warning -> error, or at least a lint. If one of the folks here who discovered this issue agrees and would like to open an issue for that, please tag me -- I'd like to keep on top of it for the sake of `cargo-semver-checks`. If nobody is interested in opening that issue, I'd be happy to do it.

Reply:
Yeah please link the issue from here once it got created. :)

Reply:
As I looks like no has opened an issue yet, I have opened https://github.com/rust-lang/rust/issues/119264